storage/innobase/handler/ha_innodb.cc

```diff
@@ -21877,6 +21906,180 @@ static void update_innodb_redo_log_encrypt(THD *thd, SYS_VAR *var,
+                                               void *save,
+                                               struct st_mysql_value *value) {
+  const char *undo_log_encryption_input;
+  char buff[4];
+  int len = sizeof(buff);
+
+  ut_a(save != NULL);
+  ut_a(value != NULL);
+
+  undo_log_encryption_input = value->val_str(value, buff, &len);
+
+  if (len > 3) return 1;
+
+  bool is_undo_log_encryption_on =
+      !innobase_strcasecmp(undo_log_encryption_input, "ON") ||
+      !innobase_strcasecmp(undo_log_encryption_input, "1");
+
+  if (is_undo_log_encryption_on == false) {
+    if (innobase_strcasecmp(undo_log_encryption_input, "OFF") != 0 &&
+        innobase_strcasecmp(undo_log_encryption_input, "0") != 0) {
+      return 1;
+    }
+  } else if (Encryption::is_online_encryption_on()) {
+    push_warning_printf(thd, Sql_condition::SL_WARNING, ER_WRONG_ARGUMENTS,
+                        " Undo log cannot be"
+                        " encrypted with Master Key encryption"
+                        " when online to KEYRING encryption is turned ON.");
+    return 1;
+  } else {
+    undo::spaces->s_lock();
+    for (auto undo_space : undo::spaces->m_spaces) {
+      fil_space_t *space = fil_space_acquire_silent(undo_space->id());
+      ut_ad(space != nullptr);
+      fil_space_crypt_t *crypt_data = space->crypt_data;
+      if (crypt_data != nullptr &&
+          crypt_data->type != CRYPT_SCHEME_UNENCRYPTED) {
+        push_warning_printf(
+            thd, Sql_condition::SL_WARNING, ER_WRONG_ARGUMENTS,
+            " Undo log cannot be"
+            " encrypted with Master Key encryption"
+            " while there are undo tablespaces encrypted with KEYRING "
+            "encryption."
+            " Please decrypt them first with encryption threads.");
+        fil_space_release(space);
+        undo::spaces->s_unlock();
+        return 1;
+      }
+      fil_space_release(space);
+    }
+    undo::spaces->s_unlock();
+  }
+
+  *reinterpret_cast<bool *>(save) = is_undo_log_encryption_on;
+  return 0;
+}
+
+static int innodb_sys_tablespace_encyption_validate(
+    THD *thd, SYS_VAR *var, void *save, struct st_mysql_value *value) {
+  const char *innodb_sys_tablespace_encryption_input;
+  char buff[STRING_BUFFER_USUAL_SIZE];
+  int len = sizeof(buff);
+
+  ut_a(save != NULL);
+  ut_a(value != NULL);
+
+  innodb_sys_tablespace_encryption_input = value->val_str(value, buff, &len);
+
+  bool legit_value = false;
+  uint use = 0;
+  for (; use < array_elements(sys_tablespace_encrypt_names); use++) {
+    if (!innobase_strcasecmp(innodb_sys_tablespace_encryption_input,
+                             sys_tablespace_encrypt_names[use])) {
+      legit_value = true;
+      break;
+    }
+  }
+
+  if (legit_value == false) return 1;
+
+  const auto change_to_sys_tablespace_encryption =
+      static_cast<srv_sys_tablespace_encrypt_enum>(use);
+
+  if (change_to_sys_tablespace_encryption == SYS_TABLESPACE_ENCRYPT_OFF) {
```

Comment:
Maybe make some 2D arrays indexed by srv_sys_tablespace_encrypt and change_to_sys_tablespace_encryption, one bool whether it's allowed, another for the warning message

Reply:
Hm, I am not sure. Good idea to be sure that all options are covered, on the other hand, there is a special case for change == ON and srv_sys_tablespace_encrypt == RE_ENCRYPT and it is a quite big one.

Reply:
if (special case) special_case(); else matrix(); ?

Reply:
special case means half of the code stays, half goes into matrix. Special case is half of existing code. But if you as a side reader would find the matrix more readable with this "half the code" :) special case then I will implement the matrix. What do you say ?

Reply:
No, all I ask you is to consider, and it looks like you have considered :) OK to leave as is